Framework/Algorithms/src/EstimateResolutionDiffraction.cpp

```diff
@@ -180,40 +209,55 @@ void EstimateResolutionDiffraction::estimateDetectorResolution() {
     }
 
     // Get the distance from detector to source
-    double l2 = spectrumInfo.l2(i);
+    const double l2 = spectrumInfo.l2(i);
 
     // Calculate T
-    double centraltof = (l1 + l2) / m_centreVelocity;
+    const double centraltof = (l1 + l2) / m_centreVelocity;
 
     // Angle
-    double twotheta =
+    const double twotheta =
         spectrumInfo.isMonitor(i) ? 0.0 : spectrumInfo.twoTheta(i);
-    double theta = 0.5 * twotheta;
+    const double theta = 0.5 * twotheta;
 
-    double solidangle = 0.0;
-    for (const auto detID : m_inputWS->getSpectrum(i).getDetectorIDs()) {
-      const auto index = detectorInfo.indexOf(detID);
-      if (!detectorInfo.isMasked(index))
-        solidangle += detectorInfo.detector(index).solidAngle(samplepos);
+    double deltatheta = 0.;
+    if (m_divergenceWS) {
+      deltatheta = m_divergenceWS->readY(i)[0];
+    } else {
+      double solidangle = 0.0;
+
+      for (const auto detID : m_inputWS->getSpectrum(i).getDetectorIDs()) {
+        const auto index = detectorInfo.indexOf(detID);
+        if (!detectorInfo.isMasked(index))
+          solidangle += detectorInfo.detector(index).solidAngle(samplepos);
+      }
+      deltatheta = sqrt(solidangle);
     }
-    double deltatheta = sqrt(solidangle);
 
     // Resolution
-    double t1 = m_deltaT / centraltof;
-    double t2 = detdim / (l1 + l2);
-    double t3 = deltatheta * (cos(theta) / sin(theta));
-
-    double resolution = sqrt(t1 * t1 + t2 * t2 + t3 * t3);
-    if (spectrumInfo.isMonitor(i))
-      resolution = 0.0;
+    const double t1 = m_deltaT / centraltof;
+    const double t2 = detdim / (l1 + l2);
+    const double t3 = deltatheta / tan(theta);
+
+    if (spectrumInfo.isMonitor(i)) {
+      m_resTof->mutableY(i)[0] = 0.;
```

Comment:
Note that you could also drop the `[0]` everywhere since `HistogramY` and similar types support assignment with a scalar. Since your workspace has just 1 point in each histogram the result should be the same.